src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -304,287 +451,468 @@ namespace nbl::ext::imgui
+					io.AddKeyEvent(bind.target, true);
+					io.AddInputCharacter(iCharacter);
 				}
-			}
+				else if (e.action == SKeyboardEvent::ECA_RELEASED)
+					io.AddKeyEvent(bind.target, false);
 		}
 	}
 
-	UI::UI(smart_refctd_ptr<ILogicalDevice> device, int maxFramesInFlight, video::IGPURenderpass* renderpass, IGPUPipelineCache* pipelineCache, smart_refctd_ptr<IWindow> window)
-		: m_device(core::smart_refctd_ptr(device)), m_window(core::smart_refctd_ptr(window))
+	UI::UI(S_CREATION_PARAMETERS&& params)
+		: m_creationParams(std::move(params))
 	{
-		createSystem();
-		struct
+		auto validateResourcesInfo = [&]() -> bool
 		{
-			struct
+			auto* pipelineLayout = m_creationParams.resources.pipelineLayout;
+
+			if (pipelineLayout) // provided? we will validate your pipeline layout to check if you declared required UI resources
 			{
-				uint8_t transfer, graphics;
-			} id;
-		} families;
+				auto validateResource = [&]<IDescriptor::E_TYPE descriptorType>(const IGPUDescriptorSetLayout* const descriptorSetLayout)
+				{
+					constexpr std::string_view typeLiteral = descriptorType == IDescriptor::E_TYPE::ET_SAMPLED_IMAGE ? "ET_SAMPLED_IMAGE" : "ET_SAMPLER",
+					ixLiteral = descriptorType == IDescriptor::E_TYPE::ET_SAMPLED_IMAGE ? "texturesBindingIx" : "samplersBindingIx";
 
-		const nbl::video::IPhysicalDevice* pDevice = device->getPhysicalDevice();
-		ILogicalDevice::SCreationParams params = {};
+					auto anyBindingCount = [&m_creationParams = m_creationParams, &log = std::as_const(typeLiteral)](const IDescriptorSetLayoutBase::CBindingRedirect* redirect) -> bool
+					{
+						if (redirect->getBindingCount())
+						{
+							m_creationParams.utilities->getLogger()->log("Provided descriptor set layout has no bindings for IDescriptor::E_TYPE::%s, you are required to provide at least single default ImGUI Font Atlas texture resource & corresponsing sampler resource!", system::ILogger::ELL_ERROR, log.data());
+							return false;
+						}
 
-		auto properties = pDevice->getQueueFamilyProperties();
+						return true;
+					};
 
-		auto requestFamilyQueueId = [&](IQueue::FAMILY_FLAGS requried, std::string_view onError)
-		{
-			uint8_t index = 0;
-			for (const auto& fProperty : properties)
-			{
-				if (fProperty.queueFlags.hasFlags(requried))
-				{
-					++params.queueParams[index].count;
+					if(!descriptorSetLayout)
+					{
+						m_creationParams.utilities->getLogger()->log("Provided descriptor set layout for IDescriptor::E_TYPE::%s is nullptr!", system::ILogger::ELL_ERROR, typeLiteral.data());
+						return false;
+					}
 
-					return index;
-				}
-				++index;
-			}
+					const auto* redirect = &descriptorSetLayout->getDescriptorRedirect(descriptorType);
 
-			logger->log(onError.data(), system::ILogger::ELL_ERROR);
-			assert(false);
-		};
+					if constexpr (descriptorType == IDescriptor::E_TYPE::ET_SAMPLED_IMAGE)
+						if (!anyBindingCount(redirect))
+							return false;
+					else
+					{
+						if (!anyBindingCount(redirect))
```

Comment:
you don't need to ask for the array size, you have the info right here!

Reply:
I ask you for ImGUI resource bindings info in creation params - I don't care if you have more resources, but you give me the layout & validate if you don't lie about what you declare

Reply:
there is a reason why decided to codegen the shaders, all the info I have in creation parameters how to do it, but I have no guarantee you provided valid pipeline layout so I check it

Reply:
> you don't need to ask for the array size, you have the info right here!

onto this one, it isn't true, if you have non-immutable samplers you count from redirect at this point is 0, thats why I give it another try to look for immutable samplers

Reply:
my point is you shouldn't need the count of array elements in the image and sampler bindings since you can look them up just how you do when you validate